File: day3/main.js
const fs = require("fs").promises

const read = async(file) => (await fs.readFile(file, "utf8")).trim()

const isDigit = (chr) => chr >= '0' && chr <= '9';
const isSymbol = (chr) => chr != '.' && (chr < '0' || chr > '9');
const isGear = (chr) => chr == '*';

const compute = (input, pred) => {
    const lines = input.trim().split("\n");
    for(let y = 0; y < lines.length; y++){
        numLoop: for(let x = 0; x < lines[y].length; x++){
            if(isDigit(lines[y][x])){
                let numStr = lines[y][x]
                let xStart = x;
                x++
                while(x < lines[y].length && isDigit(lines[y][x])){
                    numStr += lines[y][x]
                    x++
                }
                for(let y2 = Math.max(0, y-1); y2 < Math.min(lines.length-1, y+2); y2++){
                    for(let x2 = Math.max(0, xStart - 1); x2 < Math.min(lines[y2].length-1, x+1); x2++){
                        if(pred(lines[y2][x2], numStr, y2, x2)){
                            continue numLoop
                        }
                    }
                }
            }
        }
    }
}

const part1 = (input) => {
    let sum = 0;

    compute(input, (chr, numStr) => {
        if(isSymbol(chr)){
            sum += parseInt(numStr)
            return true
        }
        return false
    })

    return sum;
}

const part2 = (input) => {
    let sum = 0;
    let gearMap = new Map();

    compute(input, (chr, numStr, y2, x2) => {
        if(isGear(chr)){
            let k = y2 + "/" + x2;
            if(gearMap.has(k)){
                sum += parseInt(numStr) * parseInt(gearMap.get(k));
            }else{
                gearMap.set(k, numStr)
            }
            return true
        }
        return false
    })

    return sum;
}

const main = async () => {
	const test1 = await read("inputs/test1.txt")
	const input = await read("inputs/input.txt")
	console.log("Test 1", part1(test1))
	console.log("Part 1", part1(input))
	console.log("Test 2", part2(test1))
	console.log("Part 2", part2(input))
}
main()

File: day3/src/main.rs
use std::{fs, collections::HashMap, cmp};

fn main() {
  let test1 = fs::read_to_string("inputs/test1.txt").expect("ioerror");
  let input = fs::read_to_string("inputs/input.txt").expect("ioerror");

  println!("Test 1: {}", part1(&test1));
  println!("Part 1: {}", part1(&input));
  println!("Test 2: {}", part2(&test1));
  println!("Part 2: {}", part2(&input));
}

fn is_digit(chr: char) -> bool {
    return chr >= '0' && chr <= '9';
}
fn is_symbol(chr: char) -> bool {
    return chr != '.' && (chr < '0' || chr > '9');
}
fn is_gear(chr: char) -> bool {
    return chr == '*';
}

fn compute(input: &str, mut pred: impl FnMut(char, i32, (usize, usize)) -> bool) {
    let lines: Vec<&str> = input.split("\n").collect();

    let mut chrs: Vec<Vec<char>> = Vec::new();
    for line in lines {
        chrs.push(line.chars().collect());
    }

    let mut y = 0;
    while y < chrs.len() {
        let mut x = 0;
        'numLoop: while x < chrs[y].len() {
            if is_digit(chrs[y][x]) {
                let mut numStr = String::new();
                numStr.push(chrs[y][x]);
                let x_start = x;
                x += 1;
                while x < chrs[y].len() && is_digit(chrs[y][x]){
                    numStr.push(chrs[y][x]);
                    x += 1;
                }
                let num = numStr.parse().unwrap();
                
                let mut y2 = if y > 1 { y - 1 } else { 0 };
                while y2 < cmp::min(chrs.len() - 1, y + 2){
                    let mut x2 = if x_start > 1 { x_start - 1 } else { 0 };
                    while x2 < cmp::min(chrs[y2].len() - 1, x+1) {
                        if pred(chrs[y2][x2], num, (x2, y2)) {
                            x += 1;
                            continue 'numLoop;
                        }
                        x2 += 1;
                    }
                    y2 += 1;
                }
                 
            }
            x += 1;
        }
        y += 1;
    }
    
}

fn part1(input: &str) -> i32{
    let mut sum = 0;

    let pred = |chr: char, num: i32, coord: (usize, usize)| -> bool { 
        if is_symbol(chr) {
            sum += num;
            return true;
        }
        return false;
    };
    compute(input, pred);

    return sum;
}

fn part2(input: &str) -> i32 {
    let mut sum = 0;
    let mut gearMap: HashMap<(usize, usize), i32> = HashMap::new();

    let pred = |chr: char, num: i32, coord: (usize, usize)| -> bool { 
        if is_gear(chr) {
            if gearMap.contains_key(&coord) {
                sum += num * gearMap.get(&coord).unwrap();
            }else{
                gearMap.insert(coord, num);
            }
            return true;
        }
        return false;
    };
    compute(input, pred);

    return sum;
}